feat(providers): add n8n webhook provider

ShafatRahman opened the pull request:
## Summary

- Add an `n8n:` provider for evaluating n8n webhook workflows with templated requests, normalized outputs, tool-call metadata, and caller-scoped sessions.
- Merge current `main` (`5ffeb3321`) into the contributor branch in `03c5a8840`.
- Harden webhook operation in `afea34a86` and the audit follow-up `50b35b1c2`: safe provider IDs, real GET inputs, JSON-safe templates, nested error handling, uncached sensitive webhook calls, cancellation, and reserved prompt integrity.
- Document the provider and runnable example, including its endpoint-URL and caching boundary.

## Audit Repairs

- URL-backed routes display as stable `n8n:webhook:<fingerprint>` IDs instead of persisting or printing raw webhook URLs as execution identifiers.
- `GET` workflows receive rendered body fields as query parameters, and JSON string templates preserve prompts containing quotes or newlines.
- Nested n8n response errors are surfaced as provider errors, while falsey status fields such as `error: false` or `error: null` remain successful output.
- Webhook calls bypass Promptfoo response caching because endpoint paths/query strings and session-bearing payloads can be sensitive or stateful; this also keeps raw endpoints out of response-cache diagnostics/storage.
- Templated `{{prompt}}` values now use the actual rendered prompt even when a test also defines `vars.prompt`.
- Evaluator abort signals are forwarded to the webhook request so cancelled evaluations do not wait for the n8n timeout.
- Returned session IDs no longer create empty `metadata: {}` values; metadata is emitted only when tool calls exist.

## Validation

- `npm ci --prefer-offline --registry=https://registry.npmjs.org/` (completed; repository baseline reports 21 moderate audit findings)
- `npm run jsonSchema:generate`
- `npm run f`
- `npm run l` (exits successfully; reports an existing complexity advisory in the unchanged OpenAI factory body within `src/providers/registry.ts`)
- `npm run tsc`
- `npx vitest run test/providers/n8n.test.ts test/providers/registry.test.ts test/providers/index.test.ts --sequence.shuffle=false` (`255` passed)
- `npm run build`
- `SKIP_OG_GENERATION=true npm run build` from `site/`
- `npm test` (`721` files passed, `1` skipped; `18,344` tests passed, `10` skipped)

## End-To-End QA

Ran `npm run local -- eval ... --no-cache` against local n8n-shaped HTTP webhook fixtures:

- `eval-xgX-2026-05-25T22:45:09`: POST JSON-template request with a quote/newline prompt passed; output, session, and normalized tool metadata were validated using only the fingerprint provider ID.
- `eval-EWh-2026-05-25T22:46:00`: GET request passed, with rendered prompt and supplied session received through query parameters.
- `eval-MSj-2026-05-25T22:46:12`: nested n8n error-item response intentionally produced an evaluation error instead of accepted output.
- `eval-uoR-2026-05-26T12:17:14`: POST request returned `error: false` and passed; although test vars included `prompt: do-not-send-this-var`, the webhook received the rendered `message: "Rendered request"`, and returned session/tool metadata was asserted.

## CI Note

Fresh checks were triggered for head `50b35b1c2`, but failed jobs currently stop before checkout or test execution while GitHub downloads actions or fetches the PR merge ref (including `Your account is suspended` and unavailable `codeload.github.com` action archives). Local verification above completes the executable audit while that external GitHub gating issue remains.

ShafatRahman committed:
feat(providers): add n8n provider

ShafatRahman committed:
fix(providers): add n8n api env var and doc update

mldangelo-oai committed:
Merge origin/main into feat/n8n-provider

mldangelo-oai committed:
Merge remote-tracking branch 'origin/main' into pr-6770-n8n-provider

mldangelo-oai committed:
fix(providers): complete n8n provider sync

mldangelo-oai committed:
Merge remote-tracking branch 'origin/main' into pr-6770-n8n-provider

mldangelo-oai committed:
Merge remote-tracking branch 'origin/main' into feat/n8n-provider

# Conflicts:
#	site/docs/providers/index.md

mldangelo-oai committed:
fix(providers): harden n8n response handling

mldangelo-oai committed:
Merge remote-tracking branch 'origin/main' into codex/pr-6770-20260516

# Conflicts:
#	src/types/env.ts

mldangelo-oai committed:
Merge branch 'main' into feat/n8n-provider

mldangelo-oai committed:
Merge branch 'main' into feat/n8n-provider

mldangelo-oai committed:
Merge branch 'main' into feat/n8n-provider

mldangelo-oai committed:
Merge branch 'main' into feat/n8n-provider

mldangelo-oai committed:
Merge remote-tracking branch 'origin/main' into feat/n8n-provider

mldangelo-oai committed:
fix(providers): isolate n8n sessions and errors

mldangelo-oai committed:
docs(site): clarify n8n provider session safety

mldangelo-oai commented:
@codex review

mldangelo-oai committed:
Merge remote-tracking branch 'origin/main' into mdangelo/codex/audit-pr-6770

# Conflicts:
#	src/providers/registry.ts

mldangelo-oai committed:
fix(providers): harden n8n webhook requests

mldangelo-oai commented on src/providers/n8n.ts:
```diff
@@ -0,0 +1,499 @@
+      headers,
+    };
+
+    if (method !== 'GET') {
+      fetchOptions.body = renderedBody;
```
Fixed in `afea34a86`. GET requests now append the rendered body fields as query parameters through `buildGetUrl()`, preserving prompts and supplied session IDs. This is covered by the provider regression and the real GET eval `eval-EWh-2026-05-25T22:46:00`, where the local webhook received both values.

mldangelo-oai commented on src/providers/n8n.ts:
```diff
@@ -0,0 +1,499 @@
+        await response.deleteFromCache?.();
+        return {
+          error: `n8n webhook response error: ${String(data.error)}`,
+        };
```
Fixed in `afea34a86`. Error detection now recurses through n8n item-array and `json` response shapes and evicts successful error payloads from cache. The nested-error unit regression passes, and real eval `eval-MSj-2026-05-25T22:46:12` reports the nested workflow error rather than accepting it as output.

mldangelo-oai commented on src/providers/n8n.ts:
```diff
@@ -0,0 +1,499 @@
+      try {
+        return JSON.parse(rendered);
+      } catch {
+        return rendered;
```
Fixed in `afea34a86`. Valid JSON string templates are parsed first and rendered value-by-value, so quotes and newlines in prompts remain correctly escaped; malformed JSON-like templates now fail explicitly. A regression test and real POST eval with a quote/newline prompt both pass.

mldangelo-oai commented on src/providers/n8n.ts:
```diff
@@ -0,0 +1,499 @@
+  private providerId: string;
+
+  constructor(webhookUrl: string, options: ProviderOptions = {}) {
+    this.webhookUrl = webhookUrl;
+    this.config = (options.config as N8nProviderConfig) || {};
+
+    // Validate URL
+    if (!this.webhookUrl && !this.config.url) {
+      throw new Error('n8n provider requires a webhook URL');
+    }
+
+    this.providerId = options.id || `n8n:${this.getUrl()}`;
+  }
```
Fixed in `afea34a86`. Default and URL-routed n8n provider IDs now use a stable fingerprint and `toString()` displays that ID rather than the endpoint URL. The documentation also clarifies that configuration exports and generic debug request logs remain sensitive.

mldangelo-oai commented on src/providers/n8n.ts:
```diff
@@ -0,0 +1,499 @@
+      if (sessionId) {
+        response.sessionId = sessionId;
+      }
+    }
```
Fixed in `afea34a86`. `response.sessionId` is now populated independently, while `response.metadata` is created only when tool calls exist. The session extraction regression now asserts that a session-only response has no empty metadata object.

mldangelo-oai committed:
Merge remote-tracking branch 'origin/main' into mdangelo/codex/audit-pr-6770

mldangelo-oai committed:
fix(providers): address n8n audit gaps

mldangelo-oai commented on src/providers/n8n.ts:
```diff
@@ -0,0 +1,563 @@
+  }
+
+  if ('error' in data && data.error !== undefined) {
+    return data.error;
```
Fixed in `50b35b1c2`. `getN8nError()` now ignores falsey status values instead of rejecting successful webhook output. The regression covers `false`, `null`, an empty string, and `0`; live eval `eval-uoR-2026-05-26T12:17:14` also returned `error: false` and passed.

mldangelo-oai commented on src/providers/n8n.ts:
```diff
@@ -0,0 +1,563 @@
+    }
+  }
+
+  async callApi(prompt: string, context?: CallApiContextParams): Promise<ProviderResponse> {
```
Fixed in `50b35b1c2`. `callApi()` now accepts `CallApiOptionsParams` and forwards `abortSignal` as the request signal. A provider regression asserts the evaluator signal reaches `fetchWithCache`.

mldangelo-oai commented on src/providers/n8n.ts:
```diff
@@ -0,0 +1,563 @@
+    let latencyMs: number | undefined;
+
+    try {
+      const response = await fetchWithCache<string>(url, fetchOptions, timeout, 'text');
```
Fixed in `50b35b1c2`. n8n webhook calls now pass the cache-bypass flag to `fetchWithCache`, because tokenized endpoints and session-bearing requests are sensitive and potentially stateful. Tests assert this for token-shaped URLs and the provider docs/example now explain the response-cache boundary.

mldangelo committed:
fix(providers): close remaining n8n webhook regressions

Addresses the last three open Codex P2s plus a related P1 found during
review:

- Empty `tool_calls: []` (and `actions: []`) workflows previously
  attached `metadata.toolCalls: []` to every response because
  `if (toolCalls)` accepts empty arrays. Downstream filters that test
  `metadata?.toolCalls?.length > 0` were silently fine, but filters
  that test `Boolean(metadata?.toolCalls)` saw a false positive on
  every result. Switch to `toolCalls.length > 0`.

- Non-idempotent webhooks (POST / PATCH — POST is the default) now
  pass `maxRetries: 0` to `fetchWithCache`. The default budget of 4
  could silently re-deliver workflows that had already accepted the
  request and dispatched stateful side-effects (sending messages,
  writing to a database) before a transport-level failure surfaced.
  Idempotent methods (GET / HEAD / OPTIONS / PUT / DELETE) keep the
  default retry budget.

- Provider ID fingerprint now covers method, body, headers,
  transformResponse, sessionField, sessionHeader, and sessionParser in
  addition to the URL. Two providers wired against the same URL but
  with different request shapes now produce distinct
  `n8n:webhook:<hash>` IDs — eval pipelines key per-result by
  `provider.id()`, so a URL-only hash silently collided unrelated
  results. Hashing uses sorted-key canonical JSON (per the
  cache-key-hygiene guidance in src/providers/AGENTS.md) so semantic
  key reorderings stay stable.

Also strips basic-auth credentials and known sensitive query
parameters (api_key, token, signature, …) from URLs before the shared
`fetchWithRetries` writes them to debug / rate-limit log lines. This
is a broad benefit for every provider that uses the cache layer —
n8n is the most exposed (tokenized webhooks), but every HTTP-based
provider with credentials-in-URL was leaking the same way. Path-as-
secret URLs (`/webhook/<unguessable-id>`) still appear in debug logs
by design; the n8n docs admonition now calls this out explicitly.

Adds regression coverage for each of the four fixes and a sorted-key
canonicalization test for the fingerprint.

mldangelo committed:
Merge remote-tracking branch 'origin/main' into feat/n8n-provider

# Conflicts:
#	src/envars.ts

mldangelo committed:
fix(providers): address Codex review on n8n provider

Three new findings from the latest Codex review pass:

- **Secrets in provider fingerprint** (P1) — `getSafeProviderId` hashed
  `config.headers` raw, including rendered `Authorization: Bearer …`
  values. Per `src/providers/AGENTS.md` cache-key hygiene, secrets must
  not enter cache keys; `provider.id()` is even more visible (persisted
  to disk, shown in result JSON / web UI / debug logs). The live token
  also flipped the ID on rotation, fragmenting historical eval groupings.
  Now redacts known auth-shaped header values
  (Authorization / Cookie / Proxy-Authorization / X-Api-Key /
  X-Auth-Token / X-Access-Token / X-Secret / X-Signature / Bearer) before
  fingerprinting; header NAMES still participate so auth-vs-no-auth still
  produces distinct IDs.

- **Method case sensitivity** (P2) — `this.config.method || 'POST'` did
  not uppercase, so `method: get` in YAML left `method === 'GET'` false,
  bypassing `buildGetUrl()` and sending a body on a GET request (undici
  rejects with "Request with GET/HEAD method cannot have body"). The
  retry-policy idempotency check also missed mixed-case verbs. Now
  normalizes once at the top of `callApi` and reuses the uppercase form
  in the fingerprint too.

- **Async transformResponse Promise leak** (P2) — `parseResponse` was
  synchronous and returned the raw result of a user-supplied
  `transformResponse` function. An async function returned an unresolved
  Promise, which got placed into `ProviderResponse.output` and stringified
  to `[object Promise]` downstream. Made `parseResponse` async and
  `await`ed both the function and the `new Function`-evaluated expression;
  `callApi` now `await`s `parseResponse`.

Adds regression tests covering each:
- Rotated bearer tokens / API keys produce the same provider ID, while
  non-sensitive headers still differentiate, and auth presence vs absence
  still differentiates.
- Lowercase `method: get` builds a query-string URL, sends no body, and
  uses the default (idempotent) retry budget; mixed-case `method: Post`
  uses `maxRetries: 0`.
- Async transformResponse resolves to the awaited value, not a Promise.

mldangelo-oai committed:
Merge remote-tracking branch 'origin/main' into mdangelo/codex/review-pr6770-sync-main

# Conflicts:
#	src/providers/registry.ts

mldangelo committed:
ci: trigger CI re-run after fork-PR workflow delay

mldangelo-oai committed:
fix(providers): sanitize n8n connection diagnostics

mldangelo-oai committed:
Merge remote-tracking branch 'origin/review-pr-6770-current' into mdangelo/codex/review-pr6770-sync-main
